### PR TITLE
Crash-time journal support

### PR DESCRIPTION
**Pull request by kstenerud**
## Goal

This PR implements the crash-time journal writing API for the journaling system

	| --------------------------------------------------------------- |
	|                          Journal API                            |
	| --------------------------------------------------------------- |
	| Journal       | Thread-safety | Shared Memory | File Management |
    | ------------- |               |               |                 |
	| Document Path |               |               |                 |
	| ----------------------------- |               | --------------- |
	|   Crash-time synchronization  |               | Crash-time API  |
	| --------------------------------------------------------------- |

## Design

Everything in this part must be async-safe, so we can't use JSON libraries to implement it. Fortunately, the journal format only uses a subset of JSON.

## Testing

Added more unit tests.

### CODE REVIEW

File: bugsnag-plugin-android-ndk/src/test/cpp/main.c
Comment (kstenerud) [inline bugsnag-plugin-android-ndk/src/test/cpp/main.c:34]: Reminder: Uncomment these before merging.

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (kstenerud) [inline bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:121]: This big chunk comes directly from bugsnag-cocoa.

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (kstenerud) [inline bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:294]: End of code copied from bugsnag-cocoa

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (fractalwrench) [inline bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:36]:
```suggestion
    "request",   // BSG_CRUMB_REQUEST
```

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (fractalwrench) [inline bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:28]: This doesn't seem ideal - is there any way to enforce keeping these in sync?

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (fractalwrench) [reply bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:165]: It might be worth extracting the functions to a separate file to make the separation obvious, and reduce the source file length

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (fractalwrench) [inline bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:346]: What is the performance impact of calling `write()` each time one of these methods is called? A very long time ago the Cocoa notifier did a similar thing and introducing a buffer to avoid `write()` each time drastically improved performance. It'd be good to confirm that this isn't introducing similar issues: https://github.com/bugsnag/bugsnag-cocoa/pull/160#issuecomment-324703817

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (kstenerud) [reply bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:28]: I can't think of any, short of some kind of multi-language code generator...

`event.h` already had synchronization requirements with `BreadcrumbType.kt` before. There is also other existing code with similar issues (any place we use enums that have JVM analogues).

Comment (fractalwrench) [reply bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:29]: I guess we're stuck with this solution - at least the types of breadcrumbs should change very rarely, if at all.

Comment (kattrali) [reply bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:30]: Yeah, the types have never changed.

File: bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c
Comment (kstenerud) [reply bugsnag-plugin-android-ndk/src/main/jni/ctjournal.c:179]: I've now replaced it with buffered_writer.c

**Review by fractalwrench — COMMENTED** [review]
I'm happy with these changes - due to the changeset size it'd be nice to have a secondary review from either @lemnik or @kattrali to ensure that I've not missed anything.

**Review by kattrali — APPROVED** [review]

**Pull request merged**
